### PR TITLE
BinaryHeap: Use full sift down in .pop()

### LINKED ISSUE #29969: Tune BinaryHeap / sift down
See PR #29811

Find out in which cases the eager placing sift down (current / new version) is better than the sift down fully, sift up again algorithm.

Look at both `From` and `pop()` etc.

I intend to work on this, just write something to give datasets of various patterns and evaluate, but let's track it so that it doesn't get lost. Also, if I'm slow, anyone is super welcome to pick it up.

### PR DESCRIPTION
**Pull request by bluss**
BinaryHeap: Use full sift down in .pop()

.sift_down can either choose to compare the element on the way down (and
place it during descent), or to sift down an element fully, then sift
back up to place it.

A previous PR changed .sift_down() to the former behavior, which is much
faster for relatively small heaps and for elements that are cheap to
compare.

A benchmarking run suggested that BinaryHeap::pop() suffers
improportionally from this, and that it should use the second strategy
instead. It's logical since .pop() brings last element from the
heapified vector into index 0, it's very likely that this element will
end up at the bottom again.

Closes #29969
Previous PR #29811

### CODE REVIEW

**Comment by Gankro** [comment]
Can you post the numbers, for posterity?

**Comment by bluss** [comment]
Sure, they're here https://github.com/rust-lang/rust/issues/29969#issuecomment-164113823 and the direct link is https://gist.github.com/bluss/b179df31a8c683c087c7

I guess they could be interpreted in different ways. It's only testing two different data set sizes, so it's not exactly a well resolved picture of the asymptotics.

**Comment by Gankro** [comment]
TL;DR version for time travellers: this reduces benchmark time by 80-50% almost completely across the board, and never regresses it.

:100:

**Comment by Gankro** [comment]
TL;DR version for time travellers: 

This reduces benchmark time to be 1/2 to 1/5th for most inputs (small heaps, or cheap comparisons), but for certain inputs makes the benchmark time 2x slower (large heap of expensive comparisons).

**Comment by Gankro** [comment]
ARGH.

Ok I had it backwards. This PR is is basically regressing the latest perf of pop ops for small/simple cases in favour of asymptotic gains on large/complex cases. I'm used to integer/float keys for heaps, so this seems backwards!

CC @rust-lang/libs, what do you think we should optimize more for?

**Comment by bluss** [comment]
cc @dgrunwald

**Comment by Gankro** [comment]
We discussed this at the libs team triage today. There was some wondering if we could dynamically (or statically?) branch on anything to "guess" which strategy would be best. In particular, we could check the size of the heap? The "large heap" case seems like the more pressing one than the "complex key" one.

What do you think?

**Comment by bluss** [comment]
I'd prefer to merge this change, and put down improved BinaryHeap strategy in the issue list. I'm unlikely to set out for such a major project right now.

**Comment by Gankro** [comment]
Ok, that seems reasonable.

@bors r+
